src/Workspaces/Core/Portable/Rename/Renamer.RenameSymbolDocumentAction.cs

```diff
@@ -0,0 +1,122 @@
+
+                    solution = await RenameSymbolAsync(solution, symbol, _analysis.NewSymbolName, optionSet, cancellationToken).ConfigureAwait(false);
+                }
+
+                return solution;
+            }
+
+            private static async Task<SyntaxNode?> GetMatchingTypeDeclarationAsync(Document document, string name, CancellationToken cancellationToken)
+            {
+                var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+                var typeDeclarations = syntaxRoot.DescendantNodesAndSelf(n => !syntaxFacts.IsMethodBody(n)).Where(syntaxFacts.IsTypeDeclaration);
+                return typeDeclarations.FirstOrDefault(d => syntaxFacts.GetDisplayName(d, DisplayNameOptions.None).Equals(name, StringComparison.OrdinalIgnoreCase));
+            }
+
+            public static async Task<RenameSymbolDocumentAction?> TryCreateAsync(Document document, string newName, CancellationToken cancellationToken)
+            {
+                var analysis = await AnalyzeAsync(document, newName, cancellationToken).ConfigureAwait(false);
+
+                if (analysis.HasValue)
+                {
+                    return new RenameSymbolDocumentAction(analysis.Value, ImmutableArray<ErrorResource>.Empty);
+                }
+
+                return null;
+            }
+
+            private static async Task<AnalysisResult?> AnalyzeAsync(Document document, string newName, CancellationToken cancellationToken)
+            {
+                // TODO: Detect naming conflicts ahead of time
+                var originalSymbolName = Path.GetFileNameWithoutExtension(document.Name);
+                var matchingDeclaration = await GetMatchingTypeDeclarationAsync(document, originalSymbolName, cancellationToken).ConfigureAwait(false);
+
+                if (matchingDeclaration is object)
+                {
+                    var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                    var symbol = semanticModel.GetDeclaredSymbol(matchingDeclaration, cancellationToken);
+                    var newSymbolName = Path.GetFileNameWithoutExtension(newName);
+
+                    if (symbol is null || symbol.Name == newSymbolName)
+                    {
+                        return null;
+                    }
+
+                    return new AnalysisResult(document, newName, newSymbolName, symbol.Name);
+                }
+
+                return null;
+            }
+
+            private readonly struct AnalysisResult
+            {
+                public string OriginalDocumentName { get; }
+                public string NewDocumentName { get; }
+                public string NewSymbolName { get; }
+                public string OriginalSymbolName { get; }
```

Comment:
1. doc
2. i find the lack of symmetry here annoying :)   you have original/new, new/original...